Add some flexibility for ocean/ice output

JessicaMeixner-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
This adds functionality to have ocean/ice output frequency be separate from the atm model.   One time was created because there's an assumption in the post that these are the same.  This could be further modified to remove this assumption.  

Refs #1629 
Additional work to close above issue will be further testing and updates for GEFS.  There didn't appear to be a unique ocean/ice output task in GEFS workflow and I'm assuming we need similar updates to the GEFS workflow scripts as the GFS workflow.    This was not done in this PR as I do not think we will break GEFS as both atm & ocean/ice are the same 6 hourly frequency right now and this will allow HR3 to move on.  Happy to work with others to make sure GEFS has the same functionality in a follow-up PR later this week if desired/needed. 

Co-author: @aerorahul 

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
I ran a C768 HR3 free-forecast test on hera with 3 hourly atm output and 6 hourly ocean/ice output as desired.  See output here: /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/outputupdate/testout01 

Full CI testing is needed. 

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

JessicaMeixner-NOAA committed:
attempt at making ocn/ice output seperate from atm output

JessicaMeixner-NOAA committed:
adjust how workflow for ocn ice post fhr are being calculated

JessicaMeixner-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -33,7 +33,15 @@
+} >> diag_table_template
+
+local template=diag_table_template
+local SYEAR=${current_cycle:0:4} 
```
This is used for the atparse of diag_table.

aerorahul replied:
Please add:
```
# Disable variable not used warnings
# shellcheck disable=SC2034
```
right before the function declaration.
That will suppress this error.

JessicaMeixner-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -33,7 +33,15 @@
+
+local template=diag_table_template
+local SYEAR=${current_cycle:0:4} 
+local SMONTH=${current_cycle:4:2}
```
This is used for the atparse of diag_table.

JessicaMeixner-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -33,7 +33,15 @@
+local template=diag_table_template
+local SYEAR=${current_cycle:0:4} 
+local SMONTH=${current_cycle:4:2}
+local SDAY=${current_cycle:6:2} 
```
This is used for the atparse of diag_table.

JessicaMeixner-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -33,7 +33,15 @@
+local SYEAR=${current_cycle:0:4} 
+local SMONTH=${current_cycle:4:2}
+local SDAY=${current_cycle:6:2} 
+local CHOUR=${current_cycle:8:2}
```
This is used for the atparse of diag_table.

JessicaMeixner-NOAA commented:
@aerorahul please let me know if you have any comments/feedback/requested updates for this.  I believe this should allow us to run HR3 with the desired 3 hour atm output and 6 hour ocean/ice output + added most of it in a flexible way. 


@jiandewang  can you try this branch for a single scout run to confirm things are working as intended?

aerorahul commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -935,7 +935,10 @@ def _get_ufs_postproc_grps(cdump, config, component='atmos'):
             local_config['FHMAX_HF_GFS'] = config['FHMAX_GFS']
-            local_config['FHOUT_HF_GFS'] = config['FHOUT_GFS']
+            local_config['FHOUT_HF_GFS'] = config['FHOUT_OCNICE_GFS']
+            local_config['FHOUT_GFS']=config['FHOUT_OCNICE_GFS']
+            local_config['FHOUT_GFS']=config['FHOUT_OCNICE_GFS']
```
```suggestion
            local_config['FHOUT_GFS'] = config['FHOUT_OCNICE_GFS']
```
Removing duplicate

JessicaMeixner-NOAA replied:
removed.

JessicaMeixner-NOAA committed:
surpress warning

aerorahul commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -935,7 +935,10 @@ def _get_ufs_postproc_grps(cdump, config, component='atmos'):
+            local_config['FHOUT_HF_GFS'] = config['FHOUT_OCNICE_GFS']
+            local_config['FHOUT_GFS']=config['FHOUT_OCNICE_GFS']
+            local_config['FHOUT_GFS']=config['FHOUT_OCNICE_GFS']
+            local_config['FHOUT']=config['FHOUT_OCNICE']
```
```suggestion
            local_config['FHOUT'] = config['FHOUT_OCNICE']
```
Fix pycodestyle error

JessicaMeixner-NOAA committed:
Update workflow/rocoto/gfs_tasks.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

JessicaMeixner-NOAA committed:
fix pycode style

aerorahul commented on parm/config/gefs/config.base.emc.dyn:
```diff
@@ -220,9 +220,10 @@ export gfs_cyc=@gfs_cyc@ # 0: no GFS cycle, 1: 00Z only, 2: 00Z and 12Z only, 4:
 export FHMIN=${FHMIN_GFS}
 export FHMAX_GFS=@FHMAX_GFS@
-export FHOUT_GFS=6 # Must be 6 for S2S until #1629 is addressed; 3 for ops
+export FHOUT_GFS=6 
```
Thank you for also making parallel updates in the GEFS configs.

JessicaMeixner-NOAA replied:
I think there is still more work that's needed in workflow/rocoto/gefs_tasks.py but was a bit at a loss as to what/where this needs to be added since I didn't see ocean/ice output there

WalterKolczynski-NOAA replied:
GEFS isn't using the groups (and I hope to remove them from GFS at some point as well), so the  `_get_forecast_hours()` function in `rocoto/tasks.py` would need to be updated. Really, the modification in `_get_ufs_postproc_grps()` should be made there instead anyway (a new `component` argument will be needed, similar to the one _get_ufs_postproc_grps has).

JessicaMeixner-NOAA replied:
I'm not sure I understand what I'm being asked to do - if anything.  Please let me know if there's something specific needed from me to update.

WalterKolczynski-NOAA replied:
Opened a quick (untested) PR into this branch

WalterKolczynski-NOAA replied:
GEFS should be handled now with this change in.

JessicaMeixner-NOAA committed:
remove duplicate

WalterKolczynski-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -27,6 +27,7 @@ export FHMAX=${FHMAX_GFS}
 export FHOUT=${FHOUT_GFS}
 export FHMAX_HF=${FHMAX_HF_GFS}
 export FHOUT_HF=${FHOUT_HF_GFS}
+export FHOUT_OCNICE_GFS=${FHOUT_OCNICE}
```
This one is looks backwards (compare with existing)

JessicaMeixner-NOAA replied:
Good catch @WalterKolczynski-NOAA  - fixed.

JessicaMeixner-NOAA committed:
fix gefs config fcst

WalterKolczynski-NOAA committed:
Move component modification of output times

The forecast hour list used to create the rocoto product metatasks
were being made in the groups function instead of the forecast hour
list function.

WalterKolczynski-NOAA committed:
Fix the marine start time

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests.

aerorahul reviewed (approved):
looks good.
thanks!

JessicaMeixner-NOAA commented:
> ```
> Experiment C48_ATM_1e46fc9e  *** FAILED *** on Orion
> Experiment C48_ATM_1e46fc9e  with 1 tasks failed at 02/20/24 02:50:24 PM
> Error logs:
> /work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2327/RUNTESTS/COMROOT/C48_ATM_1e46fc9e/logs/2021032312/gfsstage_ic.log
> ```

This error is: 

```
Begin stage_ic.sh at Wed Feb 21 05:35:20 UTC 2024
+ stage_ic.sh[6]: . /work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2327/global-workflow/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
Lmod Warning: MODULEPATH is undefined.



Lmod has detected the following error: The following module(s) are unknown:
"contrib"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "contrib"

Also make sure that all modulefiles written in TCL start with the string
#%Module
```

I'll start looking into this more now.

JessicaMeixner-NOAA commented:
So I just ran a test on a local clone and I'm able to run the gfsstage_ic job.   I'm not sure what went wrong in the CI test though.  It's failed twice in the same way though, so something does appear to be going on unless something was off last night and is fixed this morning.

JessicaMeixner-NOAA commented:
@aerorahul @WalterKolczynski-NOAA any advice for what to do here?   I'm at a loss for a next step.

aerorahul commented:
I am inclined to merge this while the Orion CI failures are being investigated.
@WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA Please take note.

TerrenceMcGuinness-NOAA commented:
Note taken.
